### PR TITLE
add new dashboard

### LINKED ISSUE #217: Feature: add transparency to app.
Add dashboard, probably new metics that will increase transparency of app.

### PR DESCRIPTION
**Pull request by ansakharov**
New dashboard and metrisc for more transparency

Fixes #217

### CODE REVIEW

File: fd/file.d.go [hunk 1]
```diff
@@ -304,3 +309,51 @@ func (f *FileD) serveFreeOsMem(_ http.ResponseWriter, _ *http.Request) {
...
+	mux.HandleFunc("/pipelines/", f.servePipelines())
+}
+
+type pipeInfo struct {
+	In         pipeline.InPluginObservabilityInfo
+	Out        pipeline.OutPluginObservabilityInfo
+	LogChanges pipeline.LogChangesDTO
+}
+
+func (f *FileD) servePipelines() func(http.ResponseWriter, *http.Request) {
```
[the remark this answers is not in the document]
Comment (ansakharov) [reply fd/file.d.go:326]: brevity is good

File: pipeline/batch.go [hunk 2]
```diff
@@ -62,7 +73,7 @@ func (b *Batch) isReady() bool {
 }
 
 type Batcher struct {
-	opts BatcherOptions
+	opts *BatcherOptions
```
[the remark this answers is not in the document]
Comment (ansakharov) [reply pipeline/batch.go:76]: This recommended by linter due size of struct: 104b.
Other way: add //nolint which is worse lmao.

File: pipeline/pipeline_info.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+	result.LogChanges = changes
+	in := inObservabilityInfo{
+		PluginName: inputInfo.Type,
+	}
+	inInfo, err := p.input.GetObservabilityInfo()
+	if err != nil {
+		return pluginsObservabilityInfo{}, err
+	}
+	if inInfo.WatcherInfo.IsValid {
```
Comment (vitkovskii) [inline pipeline/pipeline_info.go:85]: pipeline shouldn't know anything about input plugin details

**Review by vitkovskii — CHANGES_REQUESTED** [review]

File: pipeline/pipeline.go [hunk 4]
```diff
@@ -193,11 +193,11 @@ func (p *Pipeline) IncMaxEventSizeExceeded() {
 }
 
 func (p *Pipeline) registerMetrics() {
-	p.inUseEventsMetric = p.metricsCtl.RegisterGauge("event_pool_in_use_events", "Count of pool events which is used for processing")
+	p.inUseEventsMetric = p.metricsCtl.RegisterGauge("event_pool_in_use_events", "FileCount of pool events which is used for processing")
```
[the remark this answers is not in the document]
Comment (ansakharov) [reply pipeline/pipeline.go]: Hooked with auth renaming.

File: fd/file.d.go [hunk 5]
```diff
@@ -54,13 +56,13 @@ func (f *FileD) Start() {
...
-	f.startHTTP()
-	f.startPipelines()
+	mux := f.startHTTP()
+	f.startPipelines(mux)
 }
 
 func (f *FileD) initMetrics() {
 	f.metricCtl = metric.New("file_d")
-	f.longPanicMetric = f.metricCtl.RegisterCounter("long_panic", "Count of panics in the LongPanic")
+	f.longPanicMetric = f.metricCtl.RegisterCounter("long_panic", "FileCount of panics in the LongPanic")
```
[the remark this answers is not in the document]
Comment (ansakharov) [reply fd/file.d.go]: Hooked it with auto renaming

File: pipeline/pipeline_info.go (hunk 3 above)
Comment (ansakharov) [reply pipeline/pipeline_info.go]: I removed this code. Now this logic hidden in plugin and html templates

File: pipeline/pipeline.go [hunk 6]
```diff
@@ -585,7 +588,25 @@ type deltas struct {
...
+	Rate             int           `json:"rate"`
+	RateMB           float64       `json:"rate_mb"`
+	ReadOps          int           `json:"read_ops"`
+	InputEvents      int64         `json:"input_events"`
+	InputTotalSize   float64       `json:"input_total_size"`
+	AvgEventSize     int64         `json:"avg_event_size"`
+	MaxEventSize     int           `json:"max_event_size"`
+}
+
+func (p *Pipeline) LogChanges(myDeltas *deltas) LogChangesDTO {
```
[the remark this answers is not in the document]
Comment (ansakharov) [reply pipeline/pipeline.go]: I renamed it to GetChanges

File: fd/file.d.go [hunk 7]
```diff
@@ -304,3 +309,51 @@ func (f *FileD) serveFreeOsMem(_ http.ResponseWriter, _ *http.Request) {
...
+	Out        pipeline.OutPluginObservabilityInfo
+	LogChanges pipeline.LogChangesDTO
+}
+
+func (f *FileD) servePipelines() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		tmpl, err := template.ParseFS(pipeline.PipelineTpl, "htmltpl/start_page.html")
+		if err != nil {
+			logger.Errorf("can't parse html template: %s", err.Error())
+			_, _ = fmt.Fprintf(w, "<hmtl><body>can't parse html: %s", err.Error())
```
[the remark this answers is not in the document]
Comment (ansakharov) [reply fd/file.d.go]: done

File: plugin/output/s3/s3_test.go [hunk 8]
```diff
@@ -9,6 +9,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/ozontech/file.d/metric"
+
```
Comment (vano144) [inline plugin/output/s3/s3_test.go:13]: ?

File: plugin/input/kafka/kafka.go [hunk 9]
```diff
@@ -171,3 +171,8 @@ func disassembleSourceID(sourceID pipeline.SourceID) (index int, partition int32
 func (p *Plugin) PassEvent(event *pipeline.Event) bool {
 	return true
 }
+
+// GetObservabilityInfo returns plugin info.
+func (p *Plugin) GetObservabilityInfo() (pipeline.InPluginObservabilityInfo, error) {
+	return pipeline.InPluginObservabilityInfo{}, nil
```
Comment (vano144) [inline plugin/input/kafka/kafka.go:177]: maybe we can create empty objects only once?

File: plugin/input/file/watcher.go [hunk 10]
```diff
@@ -18,6 +20,7 @@ type watcher struct {
 	watcherCh         chan notify.EventInfo
 	shouldWatchWrites bool
 	logger            *zap.SugaredLogger
+	fdCounter         atomic.Int64
```
Comment (vano144) [inline plugin/input/file/watcher.go:23]: where is it used?

File: plugin/action/throttle/in_memory_limiter.go [hunk 11]
```diff
@@ -38,6 +39,8 @@ func (l *inMemoryLimiter) isAllowed(event *pipeline.Event, ts time.Time) bool {
 	defer l.mu.Unlock()
 
 	id := l.rebuildBuckets(ts)
+
+	fmt.Println("bucket id", id)
```
Comment (vano144) [inline plugin/action/throttle/in_memory_limiter.go:43]: ?

File: plugin/input/file/file.go [hunk 12]
```diff
@@ -76,6 +76,13 @@ const (
 	offsetsOpReset                     // * `reset` – resets an offset to the beginning of the file
 )
 
+const (
+	// Observability consts.
+	FdCount   string = "FdCount"
```
Comment (vano144) [inline plugin/input/file/file.go:81]: FdCount = "FdCount" ?

File: fd/file.d.go [hunk 13]
```diff
@@ -304,3 +309,47 @@ func (f *FileD) serveFreeOsMem(_ http.ResponseWriter, _ *http.Request) {
 func (f *FileD) serveLiveReady(_ http.ResponseWriter, _ *http.Request) {
 	logger.Infof("live/ready OK")
 }
+
+type pipeInfo struct {
```
Comment (vano144) [inline fd/file.d.go:313]: what do you think about putting it and serve... functions in a separate file?

File: pipeline/batch.go [hunk 14]
```diff
@@ -79,8 +90,24 @@ type Batcher struct {
...
+
+	// for observability purposes
+	committedCountersMu sync.RWMutex
+	// committedCounters contains pair: timestamp - number of commits during timestamp
+	committedCounters map[int64]int64
+
+	// batcherTimeWaitMap contains two timestamps and durations of fastest and slowest committed batches during
+	// last 5 minutes. At index 0 lies fastest at index 1 lies slowest
+	batcherTimeWaitMap map[int64]BatcherTimeDTO
+	batcherTimeMu      sync.RWMutex
```
Comment (vano144) [inline pipeline/batch.go:102]: it is good practice to write fields after mutex, which are locked by its.

File: plugin/action/throttle/in_memory_limiter.go (hunk 11 above)
Comment (ansakharov) [reply plugin/action/throttle/in_memory_limiter.go:43]: forgotten debug statement

File: pipeline/batch.go (hunk 14 above)
Comment (ansakharov) [reply pipeline/batch.go:102]: thx for advice.

File: plugin/input/file/watcher.go (hunk 10 above)
Comment (ansakharov) [reply plugin/input/file/watcher.go]: it's artifacts, thx for notice. Removed.

File: plugin/input/file/file.go [hunk 15]
```diff
@@ -252,3 +262,22 @@ func (p *Plugin) PassEvent(event *pipeline.Event) bool {
...
 }
+
+// GetObservabilityInfo returns info about file plugin.
+func (p *Plugin) GetObservabilityInfo() (pipeline.InPluginObservabilityInfo, error) {
+	offsets, err := p.jobProvider.offsetDB.load()
+	if err != nil {
+		return pipeline.InPluginObservabilityInfo{}, err
+	}
+
+	Files := make(map[string]int, len(offsets))
```
Comment (vano144) [inline plugin/input/file/file.go:273]: why not just files?

File: pipeline/batch.go [hunk 16]
```diff
@@ -79,8 +90,25 @@ type Batcher struct {
 
 	outSeq    int64
 	commitSeq int64
+
+	// for observability purposes
+	// committedCounters contains pair: timestamp - number of commits during timestamp
+	committedCountersMu sync.RWMutex
+	committedCounters   map[int64]int64
```
Comment (vano144) [inline pipeline/batch.go:97]: maybe we can use here array[2] ?

File: pipeline/batch.go [hunk 17]
```diff
@@ -79,8 +90,25 @@ type Batcher struct {
...
+
+	// batcherTimeWaitMap contains two timestamps and durations of fastest and slowest committed batches during
+	// last 5 minutes. At index 0 lies fastest at index 1 lies slowest
+	batcherTimeMu      sync.RWMutex
+	batcherTimeWaitMap map[int64]BatcherTimeDTO
```
Comment (vano144) [inline pipeline/batch.go:102]: maybe we can use here array[2] ?

File: pipeline/batch.go [hunk 18]
```diff
@@ -118,19 +165,97 @@ func (b *Batcher) Start(ctx context.Context) {
...
+		}
+	}
+}
+
+func (b *Batcher) updateCommitWaitValues(ctx context.Context, batcherTimeChan <-chan BatcherTimeDTO) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case dto := <-batcherTimeChan:
```
Comment (vano144) [inline pipeline/batch.go:203]: why are we not immediately updating batcherTimeWaitMap in saveObservabilityInfo? Thus, the separate goroutine will not be needed

Comment (ansakharov) [reply pipeline/batch.go]: done.

**Pull request closed without merging**